optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,350 @@
+                :class:`~optuna.samplers.TPESampler`.
+
+        warn_independent_sampling:
+            If this is :obj:`True`, a warning message is emitted when
+            the value of a parameter is sampled by using an independent sampler.
+
+            Note that the parameters of the first trial in a study are sampled via an
+            independent sampler in most cases, so no warning messages are emitted in such cases.
+
+    Raises:
+        ValueError:
+            If ``qmc_type`` is not one of 'sobol', 'halton', 'LHS' or 'OA-LHS'.
+
+    .. note::
+        Added in v2.x.0 TODO(kstoneriv3)as an experimental feature. The interface may change in
+        newer versions without prior notice.
+
+    Example:
+
+        Optimize a simple quadratic function by using :class:`~optuna.samplers.QMCSampler`.
+
+        .. testcode::
+
+            import optuna
+
+
+            def objective(trial):
+                x = trial.suggest_float("x", -1, 1)
+                y = trial.suggest_int("y", -1, 1)
+                return x ** 2 + y
+
+
+            sampler = optuna.samplers.QMCSampler()
+            study = optuna.create_study(sampler=sampler)
+            study.optimize(objective, n_trials=20)
+
+    """
+
+    def __init__(
+        self,
+        *,
+        qmc_type: str = "sobol",
+        scramble: bool = False,
+        seed: Optional[int] = None,
+        search_space: Optional[Dict[str, BaseDistribution]] = None,
+        independent_sampler: Optional[BaseSampler] = None,
+        warn_independent_sampling: bool = True,
+        warn_asyncronous_seeding: bool = True,
+    ) -> None:
+
+        self._scramble = scramble
+        self._seed = seed or numpy.random.MT19937().random_raw()
+        self._independent_sampler = independent_sampler or optuna.samplers.RandomSampler(seed=seed)
+        self._qmc_type = qmc_type
+        self._cached_qmc_engine = None
+        # TODO(kstoneriv3): make sure that search_space is either None or valid search space.
+        # also make sure that it is OrderedDict
+        self._initial_search_space = search_space
+        self._warn_independent_sampling = warn_independent_sampling
+
+        if (seed is None) and scramble and warn_asyncronous_seeding:
+            # Sobol/Halton sequences without scrambling do not use seed.
+            self._log_asyncronous_seeding()
+
+    def reseed_rng(self) -> None:
+
+        self._independent_sampler.reseed_rng()
+        # self._seed = numpy.random.MT19937().random_raw()
+
+        # We must not reseed the self._seed. Otherwise,
+        # parallel workers will have different seed
+        # due to reseeding that happens at the begining of the parallelization.
+
+        # TODO(kstoneriv3): Maybe add some warning about reseeding because the
+        # seed of the sobol sequence will not be reseeded. However, it might be noisy
+        # in parallel execution, because reseed_rng is called when forking process.
```

Comment:
As written in the comments, there is no positive reason to use `reseed_rng` in `QMCSampler`. `reseed_rng` is used when Optuna creates parallel workers by forking processes to make sure not to sample the same random samples in different workers but in the case of QMC, the seed must be shared among workers and thus we cannot let this function update sampler's seed.

The problem is whether we should raise a warning when this method is called. If it does, the user will be faced with the warning (n_jobs - 1) times, which I think is too noisy... However, it is still important to make sure that reseed_rng does not reseed the seed of the QMC sampler itself. Maybe we can document this point... but I am not sure which is better (warning vs documentation).

Reply:
I think it's OK to raise a warning for each worker. If we use `optuna.logging.Logger.warnings()`, users can disable the warnings. (For the logging HOWTO, please check [here](https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#logging-howto) for further details.)